src/main/java/de/azapps/kafkabackup/common/offset/S3OffsetSink.java

```diff
@@ -1,27 +1,133 @@
 package de.azapps.kafkabackup.common.offset;
 
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import de.azapps.kafkabackup.storage.s3.AwsS3Service;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 import lombok.RequiredArgsConstructor;
 import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.ConsumerGroupListing;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.errors.RetriableException;
 
 @RequiredArgsConstructor
 public class S3OffsetSink implements OffsetSink {
     private final String bucketName;
     private final AdminClient adminClient;
+    private final AwsS3Service awsS3Service;
 
+    private Map<TopicPartition, OffsetStoreS3File> topicOffsets = new HashMap<>();
+    private List<String> consumerGroups;
 
     public void syncConsumerGroups() {
-
+        try {
+            consumerGroups = adminClient.listConsumerGroups()
+                .all().get().stream()
+                .map(ConsumerGroupListing::groupId)
+                .collect(Collectors.toList());
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RetriableException(e);
+        }
```

Comment:
The most important comment got lost in the review: This looks like a copy-paste from the DiskOffsetSink. So perhaps OffsetSink should be an abstract base class containing this common logic. OR the OffsetSink should be used in all cases, but have different kinds of OffsetWriters to handle just the writing (solving the problem by composition instead of inheritance) Wdyt?